flink-runtime/src/main/java/org/apache/flink/runtime/deployment/TaskDeploymentDescriptorFactory.java

```diff
@@ -0,0 +1,296 @@
+			LogicalSlot targetSlot,
+			@Nullable JobManagerTaskRestore taskRestore,
+			Collection<ResultPartitionDeploymentDescriptor> producedPartitions) throws Exception {
+		return createDeploymentDescriptor(
+			targetSlot.getTaskManagerLocation().getResourceID(),
+			targetSlot.getAllocationId(),
+			targetSlot.getPhysicalSlotNumber(),
+			taskRestore,
+			producedPartitions);
+	}
+
+	@VisibleForTesting
+	public TaskDeploymentDescriptor createDeploymentDescriptor(
+			ResourceID location,
+			AllocationID allocationID,
+			int targetSlotNumber,
+			@Nullable JobManagerTaskRestore taskRestore,
+			Collection<ResultPartitionDeploymentDescriptor> producedPartitions) throws Exception {
+		return new TaskDeploymentDescriptor(
+			jobID,
+			serializedJobInformation,
+			taskInfo,
+			executionId,
+			allocationID,
+			subtaskIndex,
+			attemptNumber,
+			targetSlotNumber,
+			taskRestore,
+			new ArrayList<>(producedPartitions),
+			createInputGateDeploymentDescriptors(location, inputEdges, subtaskIndex, lazyScheduling));
+	}
+
+	public static TaskDeploymentDescriptorFactory fromExecutionVertex(
+			ExecutionVertex executionVertex,
+			ExecutionAttemptID executionId,
+			int attemptNumber) throws ExecutionGraphException {
+		ExecutionGraph executionGraph = executionVertex.getExecutionGraph();
+		return new TaskDeploymentDescriptorFactory(
+			executionId,
+			attemptNumber,
+			getSerializedJobInformation(executionGraph),
+			getSerializedTaskInformation(executionVertex.getJobVertex().getTaskInformationOrBlobKey()),
+			executionGraph.getJobID(),
+			executionGraph.getScheduleMode().allowLazyDeployment(),
+			executionVertex.getParallelSubtaskIndex(),
+			executionVertex.getAllInputEdges());
+	}
+
+	private static MaybeOffloaded<JobInformation> getSerializedJobInformation(
+		ExecutionGraph executionGraph) {
+		Either<SerializedValue<JobInformation>, PermanentBlobKey> jobInformationOrBlobKey =
+			executionGraph.getJobInformationOrBlobKey();
+		if (jobInformationOrBlobKey.isLeft()) {
+			return new TaskDeploymentDescriptor.NonOffloaded<>(jobInformationOrBlobKey.left());
+		} else {
+			return new TaskDeploymentDescriptor.Offloaded<>(jobInformationOrBlobKey.right());
+		}
+	}
+
+	private static MaybeOffloaded<TaskInformation> getSerializedTaskInformation(
+			Either<SerializedValue<TaskInformation>,
+			PermanentBlobKey> taskInfo) {
+		return taskInfo.isLeft() ?
+			new TaskDeploymentDescriptor.NonOffloaded<>(taskInfo.left()) :
+			new TaskDeploymentDescriptor.Offloaded<>(taskInfo.right());
+	}
+
+	private static List<InputGateDeploymentDescriptor> createInputGateDeploymentDescriptors(
+			ResourceID location,
+			ExecutionEdge[][] inputEdges,
+			int subtaskIndex,
+			boolean allowLazyDeployment) throws ExecutionGraphException {
+		List<InputGateDeploymentDescriptor> consumedPartitions = new ArrayList<>(inputEdges.length);
+
+		for (ExecutionEdge[] edges : inputEdges) {
+			ShuffleDescriptor[] consumedPartitionShuffleDescriptors =
```

Comment:
Maybe in a short way `shuffleDescriptors=getShuffleDescriptors(edges, allowLazyDeployment)`

Reply:
I think we should preserve `consumedPartition` in the name to highlight the purpose of these shuffle descriptors but it can be shortened a bit, true.